suite/tests/linux/clone.c

```diff
@@ -204,14 +226,15 @@ create_thread(int (*fcn)(void *), void *arg, void **stack, bool share_sighand,
  * Currently, this supports a fcn that does not return and calls exit() on
  * its own.
  */
-int
-make_clone3_syscall(struct clone_args *clone_args, uint clone_args_size,
-                    void (*fcn)(void))
+static int
+make_clone3_syscall(void *clone_args, ulong clone_args_size, void (*fcn)(void))
 {
-    int result;
-#    ifdef X86
-#        ifdef X64
-    uint64 clone_args_size_64 = (uint64)clone_args_size;
+#ifdef SYS_clone3
+    assert(CLONE3_SYSCALL_NUM == SYS_clone3);
+#endif
+    long result;
+#ifdef X86
+#    ifdef X64
     asm volatile("mov %[sys_clone3], %%rax\n\t"
```

Comment:
In fact I remember now that we deliberately export dynamorio_syscall to tests inside drlibc so it could be used here as well to simplify this code.

Reply:
It's better to use inline asm for this `clone3` call. If we use `dynamorio_syscall`, the child thread will also return from it and will try to use some variables local to `make_clone3_syscall` (like storing the ret value in `result`). This is fragile in the `CLONE_VM` case as it may cause a crash if enough space is not left on the new stack for such accesses. Xref https://github.com/DynamoRIO/dynamorio/pull/5143#discussion_r726347743.

It was okay to use `dynamorio_syscall` in the os.c code because we do not expect the child thread to be successfully created there, so there won't ever be any such issues.

Reply:
Looking at what dr_create_client_thread() does: it uses dynamorio_clone().  Is that in drlibc?

Reply:
I suppose it doesn't use clone3 though, at least now in the current impl.

Reply:
> Looking at what dr_create_client_thread() does: it uses dynamorio_clone()

I see that `dynamorio_clone` also redirects the child thread in asm code itself, probably due to the same reason (avoiding return-to-caller for child thread) https://github.com/DynamoRIO/dynamorio/blob/75d7b48c5843d312a29f1afeea86fc6643fb7531/core/arch/x86/x86.asm#L1594

> I suppose it doesn't use clone3 though, at least now in the current impl

Right, so we cannot use it here.

Reply:
Assuming this is resolved.